manticore/core/cpu/binja.py

```diff
@@ -380,6 +381,79 @@ def decode_instruction(self, pc):
         insn.operands = self._wrap_operands(insn.operands)
         return insn
 
+    def execute(self):
+        '''
+        Decode, and execute one instruction pointed by register PC
+        '''
+        if issymbolic(self.PC):
+            raise ConcretizeRegister(self, 'PC', policy='ALL')
+
+        if not self.memory.access_ok(self.PC, 'x'):
+            raise InvalidMemoryAccess(self.PC, 'x')
+
+        self.publish('will_decode_instruction', self.PC)
+
+        insn = self.decode_instruction(self.PC)
+        self._last_pc = self.PC
+
+        self.publish('will_execute_instruction', insn)
+
+        # FIXME (theo) why just return here?
+        if insn.address != self.PC:
```

Comment:
I'm kinda confused about when this happens, but I think we should at least log when it does

Reply:
I remember being kind of confused at this recently also, but i suspect this check is not necessary anymore -- insn.address should always be == self.PC

i think this check was first introduced in this pr when we restructured when the instruction was decoded (moved it to executor) so that `state.cpu.instruction` in a hook would refer to the instruction about to be executed. https://github.com/trailofbits/manticore/pull/205/files#diff-6ad9911cdf40511ac94cf12c0ea2b36dR432

however, i'm pretty that check was there just in case; the body shouldn't ever have actually been executed in any normal operation of manticore

a little later, in the dev-events merge we reverted some of this, and no longer decode instructions in the executor, and move it back to the cpu. https://github.com/trailofbits/manticore/pull/341/files#diff-cb6130b98f84db152ec7d0078d15e4e4L689

in that pr, this code got refactored (https://github.com/trailofbits/manticore/pull/341/files#diff-6ad9911cdf40511ac94cf12c0ea2b36dR738), but i think since we reverted the original reason that check was introduced, i think it should be ok to remove this

@yan does this sound right to you?

Reply:
That does sound right; it's likely not needed anymore. If I recall correctly, it was a temporary fix to get around instruction caching issues when we separated decoding from execution.